packages/2d/src/utils/diff.ts

```diff
@@ -0,0 +1,149 @@
+interface TransformDiff<T> {
+  inserted: TransformDiffItem<T>[];
+  deleted: TransformDiffItem<T>[];
+  transformed: TransformDiffItemTransformed<T>[];
+}
+
+interface TransformDiffItem<T> {
+  before?: T;
+  beforeIdIndex: number;
+  current: T;
+  currentIndex: number;
+}
+
+interface TransformDiffItemTransformed<T> {
+  insert: boolean;
+  remove: boolean;
+  from: TransformDiffItem<T>;
+  to: TransformDiffItem<T>;
+}
+
+interface ApplyTransformInserted<T> {
+  item: TransformDiffItem<T>;
+  order: number;
+}
+
+interface ApplyTransformResult<T> {
+  inserted: ApplyTransformInserted<T>[];
+}
+
+interface Idable {
+  id: string;
+}
+
+function getIdMap<T extends Idable>(list: T[]) {
+  const map = new Map<string, TransformDiffItem<T>[]>();
+  let before: T | undefined = undefined;
+  for (const [index, current] of list.entries()) {
+    let currentArray = map.get(current.id);
```

Comment:
nit: I don't know if this is any better, but this is one line shorter and preserves immutability:
```tsx
const currentArray = map.get(current.id) ?? [];
if (!map.has(current.id)) {
    map.set(current.id, currentArray);
}
```